### PR TITLE
Issue #5133 - Improve ResourceFactory and Resource list handling

### LINKED ISSUE #5133: Improve ResourceFactory and Resource list handling
**Jetty version**
10.0.x

**Description**
While reviewing PR #5131 it was discovered that the WebAppClassLoader constructor automatically added the WebAppContext.getExtraClassPath() entries on it's own.
While at the same time WebInfConfiguration was processing the WebAppContext.getExtraClassPath() in it's own Servlet specific ways.

https://github.com/eclipse/jetty.project/blob/e117fbe8281bd43ec5601a30d39f50a09cbcef22/jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java#L195-L196

This behavior in the WebAppClassLoader constructor should be removed.

### PR DESCRIPTION
**Pull request by joakime**
* Now parsed by `WebAppContext` into `List<Resource>`
* Reintroduced `Resource.fromList()` from old PR.
* Refactored `ResourceFactory` to never return null and always throw an exception if unable to get/create/resolve the `Resource`
* `ResourceHandler.getResource(String)` cleaned up in light of Exception handling requirement
* `Resource.addPath(String)` implementations can
  never return null now

### CODE REVIEW

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java [hunk 1]
```diff
@@ -274,35 +266,4 @@ private boolean isAcceptableLibrary(File file, Set<String> pathToClassFiles)
     }
 
     private static Field _contextField;
-
```
Comment (janbartel) [inline jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java:277]: Why has this method been deleted?

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java [hunk 2]
```diff
@@ -188,8 +188,7 @@ public HttpContent getContent(String pathInContext, int maxBufferSize) throws IO
         if (_parent != null)
         {
             HttpContent httpContent = _parent.getContent(pathInContext, maxBufferSize);
-            if (httpContent != null)
-                return httpContent;
+            return httpContent;
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: This seems part of some kind of general code cleanup, and not directly related to the extraClasspath

File: jetty-server/src/main/java/org/eclipse/jetty/server/ResourceContentFactory.java [hunk 3]
```diff
@@ -51,14 +51,12 @@ public ResourceContentFactory(ResourceFactory factory, MimeTypes mimeTypes, Comp
 
     @Override
     public HttpContent getContent(String pathInContext, int maxBufferSize)
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ResourceContentFactory.java]: So where previously we had a declared IOException, this method now throws it wrapped inside an InvalidPathException instead?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java [hunk 4]
```diff
@@ -74,7 +74,7 @@ protected void notFound(HttpServletRequest request, HttpServletResponse response
             {
             }
         });
-        _resourceService.setGzipEquivalentFileExtensions(new ArrayList<>(Arrays.asList(new String[]{".svgz"})));
+        _resourceService.setGzipEquivalentFileExtensions(new ArrayList<>(Collections.singletonList(".svgz")));
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: Again, no problem with the substance of this, it's just the inclusion of this unrelated cleanup into this PR.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 5]
```diff
@@ -147,6 +148,22 @@ public ResourceCollection(String csvResources)
         return _resources;
     }
 
+    /**
+     * Sets the resource collection's resources.
+     *
+     * @param res the resources to set
+     */
+    public void setResources(List<Resource> res)
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:162]: Maybe it would be better to change _resources from an array to a List<Resource>?

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 6]
```diff
@@ -921,4 +913,97 @@ public static URL toURL(File file) throws MalformedURLException
...
+     * return the List of Resources instances it represents.
+     * <p>
+     * Supports glob references that end in {@code /*} or {@code \*}.
+     * Glob references will only iterate through the level specified and will not traverse
+     * found directories within the glob reference.
+     * </p>
+     *
+     * @param delimitedReferences the comma {@code ,} or semicolon {@code ;} delimited
+     * String of resource references.
+     * @param globDirs true if glob references return directories within the glob as well
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: Isn't this the opposite of the foregoing comment: "Glob references will only iterate through the level specified and will not traverse found directories within the glob reference"?  The intention is that when the wildcard char * is used, deep directory traversal is not supported.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 7]
```diff
@@ -921,4 +913,97 @@ public static URL toURL(File file) throws MalformedURLException
...
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: Isn't this the opposite of the foregoing comment: "Glob references will only iterate through the level specified and will not traverse found directories within the glob reference"? The intention is that when the wildcard char * is used, deep directory traversal is not supported.

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java [hunk 8]
```diff
@@ -185,15 +185,20 @@ public WebAppClassLoader(ClassLoader parent, Context context)
...
         {
-            StringTokenizer tokenizer = new StringTokenizer(extensions, ",;");
+            StringTokenizer tokenizer = new StringTokenizer(extensions, StringUtil.DEFAULT_DELIMS);
             while (tokenizer.hasMoreTokens())
             {
                 _extensions.add(tokenizer.nextToken().trim());
             }
         }
 
         if (context.getExtraClasspath() != null)
```
Comment (janbartel) [inline jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java:195]: I don't like WebAppClassLoader setting itself up from the extraClasspath:  the WebAppClassLoader is configured by the WebInfConfiguration, so I don't see why the extraClassPath shouldn't be too.

**Review by janbartel — CHANGES_REQUESTED** [review]

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 9]
```diff
@@ -293,11 +275,13 @@ public Resource addPath(String path) throws IOException
...
         }
+
         if (resources != null)
         {
             return new ResourceCollection(resources.toArray(new Resource[0]));
         }
-        return null;
+
+        throw new MalformedURLException("path does not result in Resource: " + path);
     }
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:272]: The jetty maven plugin has a use-case for turning a ResourceCollection back into a csv list - any chance you could add that onto ResourceCollection?

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java [hunk 10]
```diff
@@ -234,18 +233,26 @@ private HttpContent load(String pathInContext, Resource resource, int maxBufferS
...
+                                compressedContent = new CachedHttpContent(compressedPathInContext, compressedResource, null);
+                                CachedHttpContent added = _cache.putIfAbsent(compressedPathInContext, compressedContent);
+                                if (added != null)
+                                {
+                                    compressedContent.invalidate();
+                                    compressedContent = added;
+                                }
                             }
                         }
+                        catch (IOException e)
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: What would happen if we didn't catch here?    A 500 response probably, but that might be more appropriate that a null response?

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java [hunk 11]
```diff
@@ -279,12 +286,20 @@ private HttpContent load(String pathInContext, Resource resource, int maxBufferS
...
+                try
+                {
+                    // Is there a precompressed resource?
+                    Resource compressedResource = _factory.getResource(compressedPathInContext);
+                    if (compressedResource.exists() && compressedResource.lastModified() >= resource.lastModified() &&
+                        compressedResource.length() < resource.length())
+                        compressedContents.put(format,
+                            new ResourceHttpContent(compressedResource, _mimeTypes.getMimeByExtension(compressedPathInContext), maxBufferSize));
+                }
+                catch (IOException e)
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: again I think we should consider not catching here.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java [hunk 12]
```diff
@@ -86,9 +86,18 @@ public String getWelcomeFile(String pathInContext)
...
-            Resource welcome = getResource(welcomeInContext);
-            if (welcome != null && welcome.exists())
-                return welcomeInContext;
+            try
+            {
+                Resource welcome = getResource(welcomeInContext);
+                if (welcome.exists())
+                    return welcomeInContext;
+            }
+            catch (IOException e)
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: again let's analyse what happens without a catch here

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java [hunk 13]
```diff
@@ -157,27 +163,28 @@ public Resource getResource(String path)
...
-        }
-        catch (Exception e)
-        {
-            LOG.debug("Unable to get Resource for {}", path, e);
+            if (r != null)
+                return r;
         }
 
-        return null;
+        throw new IOException("Unable to find Resource for " + path);
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: I don't think the message is right.
This method should still return a resource EVEN IF not found, as the request might be a PUT about to create the resource.
This method should only throw if something has gone horribly wrong.
Perhaps this should be an IllegalArguementException ?

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 14]
```diff
@@ -921,4 +913,97 @@ public static URL toURL(File file) throws MalformedURLException
...
+     * @return the list of resources parsed from input string.
+     */
+    public static List<Resource> fromList(String delimitedReferences, boolean globDirs) throws IOException
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: can't say I love `delimitedReferences`.
How about `resourceList`

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 5 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:162]: I've no probs with that, but it has to be pass by value semantics and not pass by reference, so the list will be copied no matter what.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java [hunk 15]
```diff
@@ -18,17 +18,26 @@
...
+     * implementation of ResourceFactory.
+     * The provided path can be resolved
+     * against a known Resource, or can
+     * be a from-scratch Resource.
+     * </p>
      *
      * @param path The path to the resource
-     * @return The resource or null
+     * @return The resource
+     * @throws IOException if unable to create Resource
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java:40]: perhaps stress that a Resource might not exist.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResource.java [hunk 16]
```diff
@@ -290,7 +290,9 @@ public Resource addPath(String path)
         throws IOException, MalformedURLException
     {
         if (path == null)
-            return null;
+        {
+            throw new MalformedURLException("null path");
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResource.java]: `IllegalArgException`

**Review by gregw — CHANGES_REQUESTED** [review]
Mostly LGTM, just a few niggles.

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java (hunk 1 above)
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java:277]: It's been a while. I don't recall.
Obviously nothing is using it in our code. (no references to it)

The setAccessible possibly?  but that can be easily fixed with method/field modifiers.

But I bet it's more likely to be related to the `addClassPath(webappContext.getExtraClasspath())` which is additive to existing classpath from prior context (not a replacement).

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java (hunk 10 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: The catch exists for the speculative `_factory.getResource(compressedPathInContext)` and allows the next attempt to load below if not valid.

The old `_factory.getResource()` relied on null returning.

Let me see if I can find a good testcase for this.

Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: But the compressedPathInContext should not throw unless something goes horribly wrong.  It should return a resource that does not exist and we should fall through because of an exists test not because of an exception.  We REALLY don't want exceptions to be in our normal execution path!!!!

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 12 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: In the past, if `welcomeInContext` resulted in an invalid resource, it would return null. (logging the various exceptions in the `.getResource()` call.

Now, if the Resource can be created, it will be created and then tested if it exists.
Otherwise if the Resource cannot be created (bad URL syntax, invalid characters on the filesystem, etc) it throws.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 13 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: It cannot return a resource if the Resource itself cannot be created.

Some examples:

``` java
String resultA = getResource("foo#file"); // wont work on URLResource, should probably have been ecnoded.
String resultB = getResource("bar:zed"); // wont work on windows for PathResource
String resultC = getResource("WEB-INF/lib/foo.jar!/deep/resource"); // wont work on URLResource (but will work on PathResource)
```

If you insist on returning a Resource, perhaps we should instead of throwing `IOException` 
we should return `new BadResource()` instead?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 12 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: So in what circumstances would `getResource(URIUtil.addPaths(pathInContext, _welcomes[i])` throw?
Only if the pathInContext or the welcome file we badly formatted, which I think is a configuration error and should not be caught and ignored.      I think it is fine to not catch here..... unless you can think of a valid configuration of welcome files that would sometimes throw?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 13 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: You've miss understood.     I'm not saying return a resource if it is bad.  
I'm saying that "Unable to find Resource" is the wrong message for a bad resource.
I think `IllegalArgumentException` is a possibility..... hmm except that those examples you give are legal, just not for the base resource..... so IOException is probably correct.... just that the message is wrong.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 9 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:272]: how would you handle a situation where the path itself has a comma?

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:272]: We need a proper CSV parser / generator in Jetty if this is going to work reliably. (we have the start of several of them sprinkled around the code, perhaps it's time to formalize it?)

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResource.java (hunk 16 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResource.java]: How about replacing that entire thing with `Objects.requireNotNull(path)` ?

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java (hunk 8 above)
Comment (joakime) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java:195]: We have 2 use cases ...

1. A user of WebAppContext without Configuration.  Yes, this is a thing!  sometimes accidental, most of the time intentional (see spring web, spark, akka, kotlin, etc)
2. A user of WebAppContext with Configuration.

In use case 1 we need WebAppClassLoader to configure itself.
It's super odd to have the Class you told "here, use this config" not actually be the one that uses that config.

Since this is a Jetty 10 change, we should probably step back and re-evaluate the whole WebAppContext.extraClasspath role then, as per your comments, it's really a WebInfConfiguration behavior, not a WebAppClassLoader behavior.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 17]
```diff
@@ -182,56 +193,27 @@ public void setResources(Resource[] resources)
...
-        }
-
-        StringTokenizer tokenizer = new StringTokenizer(csvResources, ",;");
-        int len = tokenizer.countTokens();
-        if (len == 0)
+        if (StringUtil.isBlank(csvResources))
         {
-            throw new IllegalArgumentException("ResourceCollection@setResourcesAsCSV(String) " +
-                " argument must be a string containing one or more comma-separated resource strings.");
+            throw new IllegalArgumentException("CSV String is blank");
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: This method name is wrong.
This javadoc is wrong.
This message is wrong.

We are not doing CSV, we are doing simple delimited strings (which is not CSV).

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 6 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: No, simple glob is supported, aka only a single asterisk (`*`) which is only the level of the glob itself.

The same as doing a `ls -1 /path/to/look/at/*` (which return all entries at that path location, including the names of the directories)

A full glob support for deep directory searching would be a double asterisk (`**`).
Supporting things like `/opt/**` and returning entries like `/opt/deep/foo.txt`.
Or things like `/code/jetty/**/target/classes` which would return all entries (directories or files) that end in `/target/classes`

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: This behavior does not "traverse found directories" it just returns the directory names found at the glob level.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 14 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: I'm not sold.
To me, that parameter is not a list, and it is not resources.

It's a String.
It supports:
* 1 or more entries. (using simple delimited behaviors with comma or semicolon).
* each entry can be one of the following types:
  * a classpath reference.
  * It can be a relative path (to the working directory).
  * It can be a fully qualified path reference.
  * It can be a Java URI. (no side effects)
  * It can be a Java URL. (with side effects)
  * It can be a path reference with a simple glob suffix.
* no entry can, itself, have an unencoded comma `,` or semicolon `;` in its name.

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java (hunk 1 above)
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java:277]: The WebAppClassLoader._context is a final field.
You cannot reset the value.

This method is broken in many different ways.
Nobody is using it, otherwise we would have had many bug reports against it already.

Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/webapp/OSGiWebappClassLoader.java:277]: I had to touch it as `webappContext.getExtraClasspath()` returns a `List<Resource>` now.

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java (hunk 2 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: I had to touch `.getContent()` method, so I was checking over other's use of it.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResource.java (hunk 16 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResource.java]: sure

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 14 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: How can that arg not be a list of resources?  The method is  `fromList` not `fromStringWithDelimitersSeparatingThingsThatCanBeVariouslyResolvedToResources`?

How about:
```
List<Resource> fromString(...)
```

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java [hunk 18]
```diff
@@ -140,44 +142,51 @@ public MimeTypes getMimeTypes()
...
+        if ((r == null || !r.exists()) && path.endsWith("/jetty-dir.css"))
+            r = getStylesheet();
 
-            return r;
-        }
-        catch (Exception e)
+        if (r == null)
         {
-            LOG.debug("Unable to get Resource for {}", path, e);
+            throw new FileNotFoundException("Resource: " + path);
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:186]: This is still the wrong exception.
The situation where r==nul does not represent file not found.  I not found file will return a resource that does not exist.
r==null (if it can happen???) represents something illegal or wrong!

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 13 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java]: still not right.  See new review.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 18 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:186]: Your description of this PR says:
 > Refactored ResourceFactory to never return null and always throw an exception if unable to get/create/resolve the Resource
 > Resource.addPath(String) implementations can never return null now

So my reading of this is that r cannot be null!   So IllegalStateException would be more appropriate

**Comment by gregw** [comment]
@joakime I'm a bit lost by this PR now.   I think it needs a bit more time and considered approach.   Are there any API changes that absolutely have to be in 10.0.0, or can we bump to 10.0.1

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 6 above)
Comment (janbartel) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: How about "true to return directories in addition to files at the level of the glob"?

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 17 above)
Comment (janbartel) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: I don't think this is a big deal ... it's been this way for years without problems, so in the interests of getting 10 out, I'm fine with it as it is.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 19]
```diff
@@ -245,59 +233,54 @@ public Resource addPath(String path) throws IOException
...
-                    resource = null;
-                }
-
-                resources.add(r);
+                potentialResources.add(r);
             }
         }
 
-        if (resource != null)
+        if (potentialResources == null || potentialResources.isEmpty())
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: How can it be empty? It is only allocated if you're going to add an entry.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 20]
```diff
@@ -245,59 +233,54 @@ public Resource addPath(String path) throws IOException
...
         {
-            return resource;
+            throw new MalformedURLException("path does not result in Resource: " + path);
         }
-        if (resources != null)
+
+        if (potentialResources.size() == 1)
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: This substantially changes the semantic:  before we only ever returned null, or  the first matching resource; now, you throw instead of null (fine) but only return the first resource if there is only one, otherwise you return the whole lot. Pretty sure there's code that's not going to deal with that. Can't we leave the semantic as it was before, ie only return the first?

Also, I don't love the var name "potentialResources", what was wrong with "resources"?

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java (hunk 8 above)
Comment (janbartel) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java:195]: In the interests of getting the rest of this applied to jetty-10, I'll drop my objection.

**Review by janbartel — CHANGES_REQUESTED** [review]
A few more comments in here for you. Also there are conflicts on the ci build.

**Comment by gregw** [comment]
On reflection, the big change in this PR is the resource methods cannot return null.  That needs to be in 10.0.0 and this PR's name probably should be changed to reflect that.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 5 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:162]: Done.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java (hunk 18 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:186]: cancel that as Context.getResources can return null;

**Review by gregw — APPROVED** [review]

**Review by janbartel — APPROVED** [review]

**Comment by gregw** [comment]
@joakime can you fix conflicts and merge

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java [hunk 21]
```diff
@@ -1227,17 +1227,29 @@ public void setWarResource(Resource war)
...
+     * @throws IOException if unable to resolve the resources referenced
+     * @see #setExtraClasspath(List)
      */
-    public void setExtraClasspath(String extraClasspath)
+    public void setExtraClasspath(String extraClasspath) throws IOException
+    {
+        setExtraClasspath(Resource.fromList(extraClasspath, false, this::newResource));
+    }
+
+    public void setExtraClasspath(List<Resource> extraClasspath)
```
Comment (basil) [inline jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java:1252]: Caused #7970.
